### ISSUE TITLE
CSP header script-src nonce is not added to inline scripts

### ISSUE THREAD

**Issue by Graph2133** [OrchardCMS/OrchardCore#13389]
### Describe the bug
I have noticed that the OrchardCore.Security library does not seem to include nonce attributes in inline script tags, which can make web applications built on OrchardCore vulnerable to certain types of attacks, such as cross-site scripting (XSS). Although the "unsafe-inline" option can fix the issue, it should not be considered a solution.

### To Reproduce
- Configure a Content Security Policy (CSP) header for a web application built on OrchardCore using the OrchardCore.Security library or the UseSecurityHeaders middleware.
- Add the CSP script-src restriction "script-src 'self' 'unsafe-eval' 'nonce-{some-random-value}'". Orchard will cache that random value, so it will not be generated randomly on each request, but it will be enough to reproduce the issue.
- Check the browser console (the issue is reproducible in the latest versions of Chrome and Edge).

### Expected behavior
All inline script tags should have randomly generated nonce attributes to ensure that their content is executed only if it matches a specific nonce value provided in the CSP header.

### Actual behavior
Inline script tags do not have nonce attribute.

### Additional info
To fix the issue, I decided to create a custom security middleware. This middleware assigns security headers to the response and generates a nonce value, which I pass to my view via HttpContext items. However, I encountered a problem: I can assign the nonce value to all custom inline scripts, but all Orchard inline scripts immediately break because there is no way to pass the nonce to dynamically generated footer/header inline scripts. Is there any way to fix this issue? Is it possible to extend/decorate the ResourceManager.cs RenderHeadScript/RenderFootScript methods to include the 'nonce' from HttpContext.Items?

**Comment by Graph2133** [comment]
I am also a little bit confused about the configuration of the allowed domains for the script-src directive in the CSP header. Should I include all absolute URLs of all scripts in the header, or just CDN domains? If I go with the first option, managing a list of allowed script URLs will be a nightmare, and I'm not sure it's a good thing to send a list of URLs with all the libraries used in the app. The second option is easy to implement, but at the same time, it makes the application more vulnerable to XSS attacks. Since there are quite a lot of different libraries used in Orchard, I'm not sure which option is the best. Can you provide some guidance on this?

**Comment by sebastienros** [comment]
Do you expect `nonce-{some-random-value}` to be some liquid template that is evaluated every time?

**Comment by sarahelsaig** [comment]
I don't know about Liquid, but for Razor it would be nice if [`ScriptTagHelper`](https://github.com/OrchardCMS/OrchardCore/blob/29d6ba2aa1e5c2aaff9133968391f9bc110faeac/src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/ScriptTagHelper.cs#L12) handled an optional `nonce` attribute too. So if you wrote
```cshtml
<script at="Foot nonce>
...
</script>
```
then it would set the value of the nonce attribute to some random number and include the same in the CSP header as mentioned above.

**Comment by sebastienros** [comment]
It makes sense that we should support that, maybe at the ResourceManager level too.

Do you know if the aspnet tag helpers support a feature like this?

**Comment by Skrypt** [comment]
Here is an example of that custom tag helper in a different library than what we use.

https://github.com/andrewlock/NetEscapades.AspNetCore.SecurityHeaders/blob/master/src/NetEscapades.AspNetCore.SecurityHeaders.TagHelpers/NonceTagHelper.cs

**Comment by Skrypt** [comment]
Also made some tests here: 

![image](https://github.com/OrchardCMS/OrchardCore/assets/3228637/dd807a55-60e1-45ce-ae9d-6d350999cca4)

Results:

![image](https://github.com/OrchardCMS/OrchardCore/assets/3228637/b4318fd1-8fb4-4ea9-9304-876845073e75)

**Comment by hishamco** [comment]
As the log said the input violates the CSP directive

**Comment by sarahelsaig** [comment]
When you added `nonce-random`, did you also add it to the script block like `<script nonce="random">`? The HTML attribute must be added to each inline script block on the page otherwise you will get the error in your screenshot.

To clarify, nonce (somehow) stands for "number only used once". Although in HTML it can be any string, not just a number. Anyway, it means that [a unique value should be randomly generated for each request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_inline_script:~:text=It%20is%20important%20to%20note%2C%20this%20nonce%20value%20needs%20to%20be%20dynamically%20generated%20as%20it%20has%20to%20be%20unique%20for%20each%20HTTP%20request). Then each inline script block must have a nonce attribute with the matching value.

**Comment by Skrypt** [comment]
No, of course, it was just to show what we have and and how to make it work.
